### PR TITLE
Implement fast version of crop function in augmentations

### LINKED ISSUE #966: Slower than new PyTorch GPU transformations
## ❓ Questions and Help
Kornia's augmentations seem to be slower than PyTorch's one. The dataset I am using is CelebA.

I have:

OS: Ubuntu 20.04
GPU: 1080ti
PyTorch: 1.8.0
torchvision: 0.9.0

```python

from torchvision.io import read_image
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader
import torchvision.transforms as TT
import kornia.augmentation as KT

def timeit(func):
    def _inner(*args, **kwargs):
        start = time()
        func(*args, **kwargs)
        end = time()
        print(f"Elapsed = {end - start :.2f}s")

    return _inner

@timeit
def test_augmentation(tr, device="cuda"):
    ds = ImageFolder(root="~/Documents/Datasets/celeba", loader=read_image)
    dl = DataLoader(ds, batch_size=128, num_workers=16, pin_memory=True)

    for batch in tqdm(dl):
        x = batch[0].to(device)
        tr(x)
        continue


def test_augmentations():
    resize = 256
    crop = 224
    device = "cuda"

    augmentations = [
        nn.Sequential(
            TT.Resize((resize, resize)),
            TT.ConvertImageDtype(dtype=torch.float),
            TT.RandomVerticalFlip(),
            TT.RandomHorizontalFlip(),
            TT.RandomCrop(crop),
        ).to(device),
        nn.Sequential(
            TT.Resize((resize, resize)),
            TT.ConvertImageDtype(dtype=torch.float),
            KT.RandomVerticalFlip(),
            KT.RandomHorizontalFlip(),
            KT.RandomCrop((crop, crop)),
        ).to(device),
    ]

    list(map(test_augmentation, augmentations))


test_augmentations()
```
PyTorch's augmentations takes 15s while Kornia takes 29s. 

Best regards,

### PR DESCRIPTION
**Pull request by edgarriba**
### Description

Fixes #966 

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -884,7 +891,26 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+                y1 = int(coords[i, 0, 1])
+                y2 = int(coords[i, 3, 1]) + 1
+                out[i] = K.geometry.transform.resize(
+                    input[i:i+1, :, y1:y2, x1:x2],
+                    self.size,
+                    interpolation='bilinear',
+                    align_corners=True)
+            return out
+        else:
+            raise NotImplementedError(f"Not supported type: {self.flags['mode']}.")
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:913]: @shijianjian check if something like this would work.
And for augmentations I would give that one by the default so that gives better performance

Comment (shijianjian) [reply kornia/augmentation/augmentation.py:913]: Add tests. We are fine with it if it passed all the tests.

Comment (shijianjian) [reply kornia/augmentation/augmentation.py:913]: Also, can you provide a benchmark between the two implementations?

**Comment by edgarriba** [comment]
@shijianjian @FrancescoSaverioZuppichini https://colab.research.google.com/drive/18rdZ2pkyNArKHI17GuY1VR6J8kF_Jf2e#scrollTo=3f6zqWBh58Zt

**Review by cceyda — COMMENTED** [review]
Adds mode (='resample' or 'slice') parameter to CenterCrop & RandomCrop
Not sure how this effects speed but otherwise LGTM

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -543,6 +544,10 @@ class CenterCrop(AugmentationBase2D):
             applied to each. Default: False.
         keepdim (bool): whether to keep the output shape the same as input (True) or broadcast it
                         to the batch form (False). Default: False.
+        mode (str): The used algorithm to crop. ``slice`` will use advanced slicing to extract the tensor based
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:547]: Let's try another name. mode is taken in other functions as in affine for bilinear or nearest. Might have some confusions there. Probably like how or something.

File: test/augmentation/test_augmentation.py [hunk 3]
```diff
@@ -2126,6 +2137,21 @@ def test_pad_if_needed(self, device, dtype):
...
+        img = torch.tensor([[
+            [0., 1., 2.],
+            [3., 4., 5.],
+            [6., 7., 8.]
+        ]], device=device, dtype=dtype)
+
+        op1 = RandomCrop(size=(2, 2), mode='resample')
+        out = op1(img)
+
+        op2 = RandomCrop(size=(2, 2), mode='slice')
```
Comment (shijianjian) [inline test/augmentation/test_augmentation.py:2151]: Why not use pytest.Mark.parameterize?

File: kornia/augmentation/augmentation.py (hunk 2 above)
Comment (edgarriba) [reply kornia/augmentation/augmentation.py:547]: I'll add for now `crop_mode`

File: test/augmentation/test_augmentation.py (hunk 3 above)
Comment (edgarriba) [reply test/augmentation/test_augmentation.py:2151]: I'm comparing op1 and op2 results
